core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -302,22 +302,33 @@ class MetadataCache(brokerId: Int) extends Logging {
         }
-        updateMetadataRequest.partitionStates.asScala.foreach { info =>
-          val controllerId = updateMetadataRequest.controllerId
-          val controllerEpoch = updateMetadataRequest.controllerEpoch
-          val tp = new TopicPartition(info.topicName, info.partitionIndex)
-          if (info.leader == LeaderAndIsr.LeaderDuringDelete) {
+
+        val controllerId = updateMetadataRequest.controllerId
+        val controllerEpoch = updateMetadataRequest.controllerEpoch
+        val cachedPartitions = new mutable.ArrayBuffer[TopicPartition]
+        updateMetadataRequest.partitionStates.asScala.foreach { state =>
+          // per-partition logging here proves to be very expensive due to holding up the partition lock and the request thread
+          val tp = new TopicPartition(state.topicName, state.partitionIndex)
+          if (state.leader == LeaderAndIsr.LeaderDuringDelete) {
             removePartitionInfo(partitionStates, tp.topic, tp.partition)
-            stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
-              s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            if (stateChangeLogger.isTraceEnabled)
+              stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
+                s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
             deletedPartitions += tp
           } else {
-            addOrUpdatePartitionInfo(partitionStates, tp.topic, tp.partition, info)
-            stateChangeLogger.trace(s"Cached leader info $info for partition $tp in response to " +
-              s"UpdateMetadata request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            addOrUpdatePartitionInfo(partitionStates, tp.topic, tp.partition, state)
+            if (stateChangeLogger.isTraceEnabled)
+              stateChangeLogger.trace(s"Cached leader info $state for partition $tp in response to " +
+                s"UpdateMetadata request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            cachedPartitions += tp
           }
         }
-        metadataSnapshot = MetadataSnapshot(partitionStates, controllerId, aliveBrokers, aliveNodes)
+        stateChangeLogger.info(s"Deleted ${deletedPartitions.size} partitions from metadata cache in response to UpdateMetadata " +
+          s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId (partitions: $deletedPartitions)")
+        stateChangeLogger.info(s"Cached leader info for ${cachedPartitions.size} partitions in response to UpdateMetadata " +
+          s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId (partitions: $cachedPartitions)")
```

Comment:
Maybe we don't include the actual partitions in these two logs? Just a count or something?